File: src/server/reddit.go
package main

import (
	"fmt"
)

type RedditListing struct {
	Kind string `json:"kind"`
	Data struct {
		Children []RedditT3 `json:"children"`
	} `json:"data"`
}

type RedditT3 struct {
	Kind string `json:"kind"`
	Data struct {
		Title   string  `json:"title"`
		Url     string  `json:"url"`
		Created float64 `json:"created_utc"`
	} `json:"data"`
}

func getSubredditSite(subreddit string) (*Site, error) {
	var response RedditListing
	url := fmt.Sprintf("https://www.reddit.com/%s/top/.json?sort=top&t=week", subreddit)
	err := httpGetJSON(url, &response)
	if err != nil {
		return nil, err
	}
	links := normalizeRedditResponse(response)
	return NewSite(subreddit, links), nil
}

// normalizeRedditResponse ...
func normalizeRedditResponse(listing RedditListing) []Link {
	links := make([]Link, 0, len(listing.Data.Children))
	for _, child := range listing.Data.Children {
		links = append(links,
			Link{Title: child.Data.Title, Url: child.Data.Url, Time: int(child.Data.Created)})
	}
	return links
}

File: backend/src/SlowNews/Reddit.hs
{-# LANGUAGE DeriveGeneric     #-}
{-# LANGUAGE NamedFieldPuns    #-}
{-# LANGUAGE OverloadedStrings #-}

module SlowNews.Reddit where

import           Control.Lens  ((&), (^.))
import           Data.Aeson    (FromJSON (parseJSON), ToJSON, withObject, (.:))
import           Data.Monoid   ((<>))
import           Data.Text     (Text)
import           GHC.Generics  (Generic)
import qualified Network.Wreq  as WQ
import           SlowNews.Link (Link (Link))

data Site = Site
  { subReddit :: String
  , count     :: Maybe Int
  }
  deriving (Show, Eq, Generic)

instance FromJSON Site
instance ToJSON Site

data Body = Body
  { bodyChildren :: [RLink]
  } deriving (Show, Eq)

instance FromJSON Body where
  parseJSON = withObject "Body" $ \v -> do
    d <- v .: "data"
    Body <$> d .: "children"


data RLink = RLink
  { rlinkTitle                 :: Text
  , rlinkUrl                   :: Text
  , rlinkPermalink             :: Text
  , rlinkCreatedUtc            :: Int
  , rlinkSubredditNamePrefixed :: Text
  } deriving (Show, Eq)
instance FromJSON RLink where
  parseJSON = withObject "RLink" $ \v -> do
    d <- v .: "data"
    RLink <$> d .: "title"
          <*> d .: "url"
          <*> d .: "permalink"
          <*> d .: "created_utc"
          <*> d .: "subreddit_name_prefixed"

toLink :: RLink -> Link
toLink RLink{ rlinkTitle, rlinkUrl, rlinkPermalink, rlinkCreatedUtc, rlinkSubredditNamePrefixed } =
  Link rlinkTitle rlinkUrl metaUrl rlinkCreatedUtc rlinkSubredditNamePrefixed
  where metaUrl = "https://reddit.com" <> rlinkPermalink

fetch :: Site -> IO [Link]
fetch (Site subreddit countMaybe) = do
  r <- WQ.asJSON =<< WQ.get (url countMaybe) :: IO (WQ.Response Body)
  return $ fmap toLink $ r ^. WQ.responseBody & bodyChildren
  where
    url Nothing =
      "https://www.reddit.com/r/" ++ subreddit ++ "/top/.json?sort=top&t=week"
    url (Just count) =
      url Nothing ++ "&limit=" ++ show count
